cardano-ping/src/Network/Ping.hs

```diff
@@ -0,0 +1,445 @@
+
+import qualified Codec.CBOR.Decoding as CBOR
+import qualified Codec.CBOR.Encoding as CBOR
+import qualified Codec.CBOR.Read as CBOR
+import qualified Codec.CBOR.Write as CBOR
+import qualified Control.Monad.Class.MonadTimer as MT
+import qualified Data.ByteString.Lazy as LBS
+import qualified Data.ByteString.Lazy.Char8 as LBSC
+import qualified Data.List as L
+import qualified Network.Socket as Socket
+import qualified System.IO as IO
+
+data PingOpts = PingOpts
+  { pingOptsCount    :: !Word32
+  , pingOptsHost     :: !(Maybe String)
+  , pingOptsUnixSock :: !(Maybe String)
+  , pingOptsPort     :: !String
+  , pingOptsMagic    :: !Word32
+  , pingOptsJson     :: !Bool
+  , pingOptsQuiet    :: !Bool
+  } deriving (Eq, Show)
+
+mainnetMagic :: Word32
+mainnetMagic = 764824073
+
+handshakeNum ::  MiniProtocolNum
+handshakeNum = MiniProtocolNum 0
+
+keepaliveNum :: MiniProtocolNum
+keepaliveNum = MiniProtocolNum 8
+
+nodeToClientVersionBit :: Int
+nodeToClientVersionBit = 15
+
+data LogMsg = LogMsg ByteString
+            | LogEnd
+
+logger :: StrictTMVar IO LogMsg -> Bool -> IO ()
+logger msgQueue json = go True
+  where
+    go first = do
+      msg <- atomically $ takeTMVar  msgQueue
+      case msg of
+        LogMsg bs -> do
+          let bs' = case (json, first) of
+                (True, False)  -> LBSC.pack ",\n" <> bs
+                (True, True)   -> LBSC.pack "{ \"pongs\": [ " <> bs
+                (False, True)  -> LBSC.pack "   timestamp,                         host,                          cookie,  sample,  median,     p90,    mean,     min,     max,     std\n" <> bs
+                (False, False) -> bs
+
+          LBSC.putStr bs'
+          go False
+        LogEnd -> when json $ IO.putStrLn "] }"
+
+supportedNodeToNodeVersions :: Word32 -> [NodeVersion]
+supportedNodeToNodeVersions magic =
+  [ NodeToNodeVersionV7  magic False
+  , NodeToNodeVersionV8  magic False
+  , NodeToNodeVersionV9  magic False
+  , NodeToNodeVersionV10 magic False
+  ]
+
+supportedNodeToClientVersions :: Word32 -> [NodeVersion]
+supportedNodeToClientVersions magic =
+  [ NodeToClientVersionV9  magic
+  , NodeToClientVersionV10 magic
+  , NodeToClientVersionV11 magic
+  , NodeToClientVersionV12 magic
+  , NodeToClientVersionV13 magic
+  , NodeToClientVersionV14 magic
+  ]
+
+data NodeVersion
+  = NodeToClientVersionV9  !Word32
+  | NodeToClientVersionV10 !Word32
+  | NodeToClientVersionV11 !Word32
+  | NodeToClientVersionV12 !Word32
+  | NodeToClientVersionV13 !Word32
+  | NodeToClientVersionV14 !Word32
+  | NodeToNodeVersionV1    !Word32
+  | NodeToNodeVersionV2    !Word32
+  | NodeToNodeVersionV3    !Word32
+  | NodeToNodeVersionV4    !Word32 !Bool
+  | NodeToNodeVersionV5    !Word32 !Bool
+  | NodeToNodeVersionV6    !Word32 !Bool
+  | NodeToNodeVersionV7    !Word32 !Bool
+  | NodeToNodeVersionV8    !Word32 !Bool
+  | NodeToNodeVersionV9    !Word32 !Bool
+  | NodeToNodeVersionV10   !Word32 !Bool
+  deriving (Eq, Ord, Show)
```

Comment:
I don't think these bangs are necessary, we print it to the console so it will be evaluated.

Reply:
Removed all bangs in types